### PR TITLE
Uncurrier phase

### PR DESCRIPTION
**Pull request by jmittert**
The uncurrier transforms curried function definitions into uncurried ones which
take multiple arguments.

 For example, consider
```
def add(x: Int): Int -> Int = {
   z -> x + z
}

def a: Int = (add(3))(4)
```

The uncurrier will create a function
```
def add$uncurried(x: Int, z: Int): Int = {
  x + z
}
```
And change the call site to
```
def a: Int = add$uncurried(3,4)
```

Right now it only does function definitions and their call sites. It might not be difficult to also do it for some trivial lambda cases, such as where the lambda is called where it is defined, (e.g. `(x -> x +1)(2)`).

Also, right now, it will generate the uncurried versions regardless of whether or not they are called. It probably wouldn't be difficult to keep track of which definitions are used, then remove the unused ones, but I'm not sure if we already eliminate dead code/unused functions in a later phase anyways.

### CODE REVIEW

**Comment by jmittert** [comment]
Here's the performance graph, running against FloydWarshall.flix, the phase took 2ms
image

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+      val curriedDef = pair._2
+      curriedDef.exp match {
+        // If the body is a lambda, then we have a curried function
+        // definition, so we need to create an uncurried version.
+        // To do this, we'll take make the body of the function the body
+        // of the lambda, then add the arguments of the lambda to the
+        // formal parameters
+        case Lambda(args, body, _, _) =>
+          // Create a new definition for our function
+          val uncurriedSym = Symbol.mkDefnSym(pair._1.name + "$uncurried")
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: I believe there is a "better" `mkDefnSym` to use here (which takes the original symbol as argument).

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    /**
+      * Creates all the uncurried versions of a function
+      *
+      * @param pair The dfnsym definition pair to base the functions on
+      * @return A map containing all the new definitions
+      */
+    def mkUncurriedDef(pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      val sym = pair._1
+      val curriedDef = pair._2
+      curriedDef.exp match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: This does not seem recursive to me. Does this create uncurried versions?

E.g. x -> y -> z -> w  should probably have the versions:

(x, y) -> z -> w
(x, y, z) -> w

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala (hunk 1 above)
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: In fact, I don't think you can safely create symbols this way.

Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: You will need to do this (its sucks, but it is more safe)

Use `def freshDefnSym(sym: DefnSym)(implicit genSym: GenSym)` in `Symbol`.

Maintain a map, probably of this structure:

`Map[Symbol.DefnSym, Map[Int, Symbol.DefnSym]]` which given a symbol and an arity returns the appropriate uncurried function (if it exists).

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    }
+  }
+
+  object Expressions {
+    /**
+      * Recursively replace all the variables in tast with new variables using the
+      * pairs in replaceSyms. This ensure that the new functions created don't
+      * reference the variables of the original functions.
+      *
+      * @param oldExp The expression to replace
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: Skip the documentation for parameters. Just focus on the documentation for the overall function.

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+      * @param replaceSyms The v
+      * @return
+      */
+    def replaceVars(oldExp: SimplifiedAst.Expression, replaceSyms : Map[Symbol.VarSym, Symbol.VarSym])(implicit genSym: GenSym): SimplifiedAst.Expression =  {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: or an alternative name might be `alphaRename` or `copy`.

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+      }
+    }
+
+    /**
+      * Uncurry an expression
+      *
+      * @param tast The expression to uncurry
+      * @return The expression modified to have all curried calls uncurried
+      */
+    def uncurry(tast: SimplifiedAst.Expression, root: SimplifiedAst.Root): SimplifiedAst.Expression = tast match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: As far as I can tell, this function is not recursive, but applications should be replaced inside subexpressions. E.g. if you have `let f = 42; g(1)(2)(3)` we want to replace g.

File: main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+              // Only call the uncurried if it exists
+              val sym = Symbol.mkDefnSym(sym2.name + "$uncurried")
+              if (root.definitions.contains(sym)) {
+                Apply(Ref(sym, uncurryType(tpe2), loc2), args1 ::: args0, tpe0, loc0)
+              } else {
+                tast
+              }
+            case _ => Apply(inner, args0, tpe0, loc0)
+          }
+          case _ => tast
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala]: We don't use wildcards, because we want to get warnings when the AST changes.

**Comment by magnus-madsen** [comment]
I added some comments. Overall it looks good to me, but there are some changes to be made. 

I changed the AST structure, so I apologize for the breakage.

**Comment by magnus-madsen** [comment]
Nice! Now I just need to make some changes to the frontend s.t. every function is curried.

**Pull request merged**